cpp/src/arrow/compute/exec_internal.h

```diff
@@ -72,6 +72,56 @@ class ARROW_EXPORT ExecBatchIterator {
   int64_t max_chunksize_;
 };
 
+/// \brief Break std::vector<Datum> into a sequence of non-owning
+/// ExecSpan for kernel execution. The lifetime of the Datum vector
+/// must be longer than the lifetime of this object
+class ARROW_EXPORT ExecSpanIterator {
+ public:
+  ExecSpanIterator() = default;
+
+  /// \brief Initialize itertor iterator and do basic argument validation
+  ///
+  /// \param[in] batch the input ExecBatch
+  /// \param[in] max_chunksize the maximum length of each ExecSpan. Depending
+  /// on the chunk layout of ChunkedArray.
```

Comment:
The missing param description for `output_shape` will probably error on some compilers?